src/DotNetty.Common/Utilities/HashedWheelTimer.cs

```diff
@@ -0,0 +1,712 @@
+            // we have not yet shutdown then we want to make sure we decrement the active instance count.
+            if (Interlocked.Exchange(ref this.workerStateVolatile, WorkerStateShutdown) != WorkerStateShutdown)
+            {
+                Interlocked.Decrement(ref instanceCounter);
+            }
+        }
+
+        internal CancellationToken CancellationToken => this.cancellationTokenSource.Token;
+
+        PreciseTimeSpan StartTime
+        {
+            get { return PreciseTimeSpan.FromTicks(Volatile.Read(ref this.startTimeVolatile)); }
+            set { Volatile.Write(ref this.startTimeVolatile, value.Ticks); }
+        }
+
+        static HashedWheelBucket[] CreateWheel(int ticksPerWheel)
+        {
+            if (ticksPerWheel <= 0)
+            {
+                throw new ArgumentException(
+                    "ticksPerWheel must be greater than 0: " + ticksPerWheel);
+            }
+            if (ticksPerWheel > 1073741824)
+            {
+                throw new ArgumentException(
+                    "ticksPerWheel may not be greater than 2^30: " + ticksPerWheel);
+            }
+
+            ticksPerWheel = NormalizeTicksPerWheel(ticksPerWheel);
+            var wheel = new HashedWheelBucket[ticksPerWheel];
+            for (int i = 0; i < wheel.Length; i++)
+            {
+                wheel[i] = new HashedWheelBucket();
+            }
+            return wheel;
+        }
+
+        static int NormalizeTicksPerWheel(int ticksPerWheel)
+        {
+            int normalizedTicksPerWheel = 1;
+            while (normalizedTicksPerWheel < ticksPerWheel)
+            {
+                normalizedTicksPerWheel <<= 1;
+            }
+            return normalizedTicksPerWheel;
+        }
+
+        /// <summary>
+        /// Starts the background thread explicitly. The background thread will
+        /// start automatically on demand even if you did not call this method.
+        /// </summary>
+        /// <exception cref="InvalidOperationException">if this timer has been
+        /// stopped already.</exception>
+        public void Start()
+        {
+            switch (Volatile.Read(ref this.workerStateVolatile))
+            {
+                case WorkerStateInit:
+                    if (Interlocked.CompareExchange(ref this.workerStateVolatile, WorkerStateStarted, WorkerStateInit) == WorkerStateInit)
+                    {
+                        this.workerThread.Start();
+                    }
+                    break;
+                case WorkerStateStarted:
+                    break;
+                case WorkerStateShutdown:
+                    throw new InvalidOperationException("cannot be started once stopped");
+                default:
+                    throw new InvalidOperationException("Invalid WorkerState");
+            }
+
+            // Wait until the startTime is initialized by the worker.
+            while (this.StartTime == PreciseTimeSpan.Zero)
```

Comment:
makes no sense on a second look

---
In reply to: [114474586](https://github.com/Azure/DotNetty/pull/242#discussion_r114474586) [](ancestors = 114474586)